File: apps/backend/src/modules/property-management/eventHandler.ts
import { COMPLIANCE_EVENTS } from "shared/events/modules/compliance.ts";
import type { EventEnvelope } from "shared/types/event.ts";
import type { PropertyParentDocument } from "shared/types/modules/compliance.ts";
import type { Server as SocketServer } from "socket.io";
import type { IPropertyManagementSvc } from "./svc.ts";
import type { RedisClientType } from "../../db/redis-client.ts";
import { catchErrorSync, formatAggregateError } from "shared/utils/error.ts";
import type { IReportingClient } from "./client.ts";
import { PROPERTY_MANAGEMENT_EVENTS } from "shared/events/modules/property-management.ts";
import { type Logger } from "winston";

export interface IPropertyManagementEventHandler {
  handlePropertyParentDocumentRequestApprovedEvent(): Promise<void>;
}

export class PropertyManagementEventHandler
  implements IPropertyManagementEventHandler {
  private svc: IPropertyManagementSvc;
  private reportingClient: IReportingClient;
  private redisSubscriber: RedisClientType;
  private socketServer: SocketServer;
  private logger: Logger;
  constructor(
    svc: IPropertyManagementSvc,
    reportingClient: IReportingClient,
    redisSubscriber: RedisClientType,
    socketServer: SocketServer,
    logger: Logger,
  ) {
    this.svc = svc;
    this.reportingClient = reportingClient;
    this.redisSubscriber = redisSubscriber;
    this.socketServer = socketServer;
    this.handlePropertyParentDocumentRequestApprovedEvent();
    this.logger = logger.child({
      service: "property-management-event-handler",
    });
  }

  async handlePropertyParentDocumentRequestApprovedEvent(): Promise<void> {
    this.redisSubscriber.subscribe(
      COMPLIANCE_EVENTS.PROPERTY_PARENT_DOCUMENT_APPROVED,
      async (message: string) => {
        //TODO: idempotent consumer outbox table
        const { data, error: parseError } = catchErrorSync(() =>
          JSON.parse(message),
        );
        if (parseError) {
          this.logger.error(parseError);
          return;
        }
        const event: EventEnvelope<PropertyParentDocument> = data;
        this.logger.log(
          "info",
          `Received event:\n${JSON.stringify(event, null, 2)}`,
        );

        const { data: documentType, error: documentTypeError } =
          await this.reportingClient.getParentDocumentTypeByDocumentId(
            event.payload.parent_document_id,
          );
        if (documentTypeError) {
          this.logger.error(documentTypeError);
          return;
        }
        if (!documentType) {
          this.logger.error(
            new Error(
              `No document type found for: ${event.payload.parent_document_id}!`,
            ),
          );
          return;
        }

        const promiseResults = await Promise.all([
          this.svc.getPointValueForGivenPropertyParentDocumentByDocumentType(
            event.payload.property_id,
            documentType,
          ),
          this.svc.getAllPropertyChildrenForGivenParent(
            event.payload.property_id,
            event.payload.user_id,
          ),
        ]);

        const errors = promiseResults
          .map((result) => result.error?.message)
          .filter((error) => error !== undefined);
        if (errors.length > 0) {
          const errMsg = formatAggregateError(errors);
          this.logger.error(new Error(errMsg));
          return;
        }
        const [pointValueResult, propertyChildrenResult] = promiseResults;
        const pointValue = pointValueResult.data ?? 0;
        const propertyChildren = propertyChildrenResult.data ?? [];

        const childrenIds = propertyChildren?.map((x) => x.child_id);
        const { data: incrementData, error: incrementError } =
          await this.svc.incrementPropertyChildrenPointsForGivenParent(
            event.payload.property_id,
            event.payload.user_id,
            childrenIds ?? [],
            pointValue,
          );
        if (incrementError) {
          this.logger.error(incrementError);
          return;
        }
        // Emit to socket listeners only if all above pass
        this.socketServer?.emit(
          COMPLIANCE_EVENTS.PROPERTY_PARENT_DOCUMENT_APPROVED,
          event,
        );
        this.socketServer.emit(
          PROPERTY_MANAGEMENT_EVENTS.PROPERTY_CHILDREN_UPDATED,
          incrementData,
        );
      },
    );
  }
}

File: apps/backend-python/src/modules/property_management/event_handler.py
import asyncio
from typing import List, Set
from uuid import UUID

from src.modules.property_management.client import IReportingClient
from src.modules.property_management.svc import IPropertyManagementSvc

import redis.asyncio as redis
from src.shared.handlers.event_handler import EventHandler
from src.shared.types.modules.compliance.event import COMPLIANCE_EVENT
from src.shared.types.modules.compliance.model import PropertyParentDocument
from src.shared.types.modules.property_management.model import PropertyChild
from src.shared.types.response import HTTPErrorResponse


class PropertyManagementEventHandler(EventHandler):
    EVENT_MAP = {
        COMPLIANCE_EVENT.PROPERTY_PARENT_DOCUMENT_APPROVED: PropertyParentDocument
    }

    def __init__(
        self,
        svc: IPropertyManagementSvc,
        reporting_client: IReportingClient,
        redis_client: redis.Redis,
    ) -> None:
        super().__init__(redis_client)
        self.svc = svc
        self.reporting_client = reporting_client

    @EventHandler.handle.register
    async def _(self, event: PropertyParentDocument):
        (
            document_type,
            error,
        ) = await self.reporting_client.get_parent_document_type_by_document_id(
            parent_document_id=event.parent_document_id
        )
        if error:
            self.logger.error(error)
            return
        assert document_type is not None
        tasks = await asyncio.gather(
            self.svc.get_point_value_for_given_property_parent_document_by_document_type(
                event.property_id,
                document_type,
            ),
            self.svc.get_all_property_children_for_given_parent(
                event.property_id,
                event.user_id,
            ),
        )
        points_task: HTTPErrorResponse[int] = tasks[0]
        points, error = points_task
        if error:
            self.logger.error(error)
            return
        assert points is not None
        property_children_task: HTTPErrorResponse[List[PropertyChild]] = tasks[1]
        property_children, error = property_children_task
        if error:
            self.logger.error(error)
            return
        assert property_children is not None
        children_ids: Set[UUID] = set(pc.child_id for pc in property_children)
        (
            _,
            error,
        ) = await self.svc.increment_property_children_points_for_given_parent(
            property_id=event.property_id,
            point_value=points,
            children_ids=children_ids,
        )
        if error:
            self.logger.error(error)
            return
        # TODO: socket event: COMPLIANCE_EVENTS.PROPERTY_PARENT_DOCUMENT_APPROVED
        # TODO: socket event: PROPERTY_MANAGEMENT_EVENTS.PROPERTY_CHILDREN_UPDATED
